Src/ILGPU.Algorithms/Runtime/Cuda/API/CuFFTAPI.cs

```diff
@@ -0,0 +1,672 @@
+// ---------------------------------------------------------------------------------------
+
+using System;
+
+namespace ILGPU.Runtime.Cuda.API
+{
+    /// <summary>
+    /// An implementation of the cuFFT API.
+    /// </summary>
+    [CLSCompliant(false)]
+    public abstract partial class CuFFTAPI
+    {
+        #region Static
+
+        /// <summary>
+        /// Creates a new API wrapper.
+        /// </summary>
+        /// <param name="version">The cuFFT version to use.</param>
+        /// <returns>The created API wrapper.</returns>
+        public static CuFFTAPI Create(CuFFTAPIVersion? version) =>
+            version.HasValue
+            ? CreateInternal(version.Value)
+            : CreateLatest();
+
+        /// <summary>
+        /// Creates a new API wrapper using the latest installed version.
+        /// </summary>
+        /// <returns>The created API wrapper.</returns>
+        private static CuFFTAPI CreateLatest()
+        {
+            Exception firstException = null;
+            var versions = Enum.GetValues(typeof(CuFFTAPIVersion));
+
+            for (var i = versions.Length - 1; i >= 0; i--)
+            {
+                var version = (CuFFTAPIVersion)versions.GetValue(i);
+                var api = CreateInternal(version);
+                if (api is null)
+                    continue;
+
+                try
+                {
+                    var status = api.GetVersion(out _);
+                    if (status == CuFFTResult.CUFFT_SUCCESS)
+                        return api;
+                }
+                catch (Exception ex) when (
+                    ex is DllNotFoundException ||
+                    ex is EntryPointNotFoundException)
+                {
+                    firstException ??= ex;
+                }
+            }
+
+            throw firstException ?? new DllNotFoundException(nameof(CuFFTAPI));
+        }
+
+        #endregion
+
+        #region Basic Plans
+
+        /// <summary>
+        /// Creates a custom plan.
+        /// </summary>
+        /// <param name="plan">Filled in with the created plan.</param>
+        /// <param name="rank">The transform.</param>
+        /// <param name="n">The transform dimensions.</param>
+        /// <param name="inembed">The storage dimensions of the input data.</param>
+        /// <param name="istride">The stride of the input data.</param>
+        /// <param name="idist">The distance of the input data.</param>
+        /// <param name="onembed">The storage dimensions of the output data.</param>
+        /// <param name="ostride">The stride of the output data.</param>
+        /// <param name="odist">The distance of the output data.</param>
+        /// <param name="type">The transform type.</param>
+        /// <param name="batch">The number of transforms.</param>
+        /// <returns>The error code.</returns>
+        public unsafe CuFFTResult PlanMany(
+            out IntPtr plan,
+            int rank,
+            int[] n,
```

Comment:
Question: should we change this to span types?

Reply:
Done.